### PR TITLE
[Android] Load System's OpenSSL

### PR DESCRIPTION
**Pull request by EgorBo**
We can't use `dlopen` for system libs (AFAIK since some more or less recent API level) e.g. for system OpenSSL, but it turns out it's still always loaded into the current process (it's loaded in Zygote and Android just forks it every time) because most of the Android Security still relies on OpenSSL(BoringSSL) so we actually can just `dsym` it for `dlopen(NULL, RTLD_NOW) /*open self*/` as far as I understand (if `dlopen(NULL)` works reliably). It doesn't look like it's permitted (but makes sense to try to submit a simple app into the Google Play Store for tests - it should validate an apk for internal/prohibited APIs usages). I tried locally and managed to run simple hash functions just fine without any external OpenSSL bits. The only problem - among 348 symbols used by .NET there are 46 missing in the BoringSSL:
```
CRYPTO_add_lock **** (see https://boringssl.googlesource.com/boringssl/+/HEAD/PORTING.md#reference-counts-and-opaque-types)
d2i_OCSP_RESPONSE
DSA_OpenSSL
EC_GFp_mont_method
EC_GFp_simple_method
EC_GROUP_check
EC_GROUP_get0_seed
EC_GROUP_get_seed_len
EC_GROUP_new
EC_GROUP_set_curve_GFp
EC_GROUP_set_seed
EVP_aes_128_ccm
EVP_aes_128_cfb128
EVP_aes_128_cfb8
EVP_aes_192_ccm
EVP_aes_192_cfb128
EVP_aes_192_cfb8
EVP_aes_256_ccm
EVP_aes_256_cfb128
EVP_aes_256_cfb8
EVP_des_cfb8
EVP_des_ede3_cfb8
EVP_des_ede3_cfb64
EVP_rc2_ecb
i2d_OCSP_REQUEST
i2d_OCSP_RESPONSE
OCSP_BASICRESP_free
OCSP_basic_verify
OCSP_CERTID_free
OCSP_cert_to_id
OCSP_check_nonce
OCSP_request_add0_id
OCSP_request_add1_nonce
OCSP_REQUEST_free
OCSP_REQUEST_new
OCSP_resp_find_status
OCSP_response_get1_basic
OCSP_RESPONSE_free
OCSP_RESPONSE_new
RSA_get_method
RSA_PKCS1_SSLeay
RSA_set_method
SSL_ctrl
SSL_CTX_ctrl
```
(see https://github.com/google/conscrypt/blob/master/common/src/jni/main/cpp/conscrypt/native_crypto.cc)

I believe these can be workaround-ed but I am not a System.Cryptography expert. The question is - can we rely on it?

/cc @marek-safar @akoeplinger @steveisok  @jonpryor also 'System.Security' label guys.

### CODE REVIEW

**Comment by steveisok** [comment]
My concern would be the potential of playing whack-a-mole with every release.  Or if they somehow make system OpenSSL completely inaccessible.

File: src/libraries/Native/Unix/CMakeLists.txt [hunk 1]
```diff
@@ -238,10 +238,7 @@ elseif(CLR_CMAKE_TARGET_TVOS)
     # it is only used for interacting with OpenSSL which isn't useful there
 elseif(CLR_CMAKE_TARGET_ANDROID AND NOT CROSS_ROOTFS)
     #add_subdirectory(System.Net.Security.Native) # TODO: reenable
```
Comment (bartonjs) [inline src/libraries/Native/Unix/CMakeLists.txt:240]: This line can probably be deleted now :smile:

**Comment by bartonjs** [comment]
> My concern would be the potential of playing whack-a-mole with every release.

Agreed.

This is only for 5.0, right? (Since 6 should be using Java interop)  How much crypto was enabled for Android on 5?

* If certs aren't supported then none of the OCSP stuff matters (and I guess we'd need something like REQUIRED_FUNCTION_EXCEPT_ANDROID to paint that and allow for the difference between the main and Android shims).
* If symmetric encryption isn't supported then none of the listed missing EVP_ stuff matters.
* If ECDsa/ECDiffieHellman aren't supported then none of the EC_ stuff matters.

CRYPTO_add_lock seems kinda important, though.

**Comment by EgorBo** [comment]
> My concern would be the potential of playing whack-a-mole with every release. Or if they somehow make system OpenSSL completely inaccessible.

I agree so we have 3 options at the moment:
1) Prepare a nuget with OpenSSL bits: we can set up a bot that will wrap [google's openssl](https://mvnrepository.com/artifact/com.android.ndk.thirdparty/openssl) into a nuget automatically once they release a new version.
I know we want to avoid it, but it turns out Google is able to scan apps for out-of-date openssl bits and automatically sends a notice to update them to publishers in case of known vulnerabilities. Users will just have to update the nuget and re-submit apks.

2) Use system OpenSSL (this PR) if it's legal.

3) Implement System.Security via Android SDK (in progress)
There are two options here:
    * Implement `Interop.*`-level calls with Android SDK + JNI (see https://github.com/dotnet/runtime/pull/45870/files, https://github.com/dotnet/runtime/pull/44989/files and https://github.com/dotnet/runtime/pull/43740/files) - it worked well so far for hashing and some simple algorithms like AES/RSA but I start to notice calls I have no idea how to implement at all, or some of them require access to "internal" java classes.
    * Implement high-level APIs instead of `Interop.*` - can't even estimate the efforts for this one especially without solid System.Security background.

**Comment by EgorBo** [comment]
> * If symmetric encryption isn't supported then none of the listed missing EVP_ stuff matters.

Symmetric Encryption seems already works with Java Interop so can be skipped indeed.

> CRYPTO_add_lock seems kinda important, though.

Ouch, will investigate what google uses in its Crypto

**Comment by bartonjs** [comment]
> but I start to notice calls I have no idea how to implement at all, or some of them require access to "internal" java classes.

Feel free to poke me on any of those.  My guess is that it's just places where we're too close to the OpenSSL API inside the shim (an original architectural goal that we started moving away from, especially once we were supporting two different OpenSSL API versions and needed to start abstracting more) and that we just need to back up and look at things from a higher level.

**Comment by EgorBo** [comment]
> > but I start to notice calls I have no idea how to implement at all, or some of them require access to "internal" java classes.
> 
> Feel free to poke me on any of those. My guess is that it's just places where we're too close to the OpenSSL API inside the shim (an original architectural goal that we started moving away from, especially once we were supporting two different OpenSSL API versions and needed to start abstracting more) and that we just need to back up and look at things from a higher level.

From a quick glance Android SDK doesn't even have `ASN.1` I only found an internal class I can invoke (Play Store's verifier might not like it) or `PKCS12Attribute` (requires API Level 26 or higher and it still doesn't cover all things we need).

**Comment by bartonjs** [comment]
> From a quick glance Android SDK doesn't even have ASN.1

We have System.Formats.Asn1, shouldn't need it from the native layer (places where we're letting OpenSSL export current objects using their encoder probably just need to be abstracted a bit differently).

> or PKCS12Attribute

On Linux we use a managed PKCS#12/PFX loader now, so that's probably dead code.

**Comment by EgorBo** [comment]
> > From a quick glance Android SDK doesn't even have ASN.1
> 
> We have System.Formats.Asn1, shouldn't need it from the native layer (places where we're letting OpenSSL export current objects using their encoder probably just need to be abstracted a bit differently).
> 
> > or PKCS12Attribute
> 
> On Linux we use a managed PKCS#12/PFX loader now, so that's probably dead code.

I copy-pasted some C# example for RSA and it contained `CryptoConfig.MapNameToOID("SHA1")` which uses two ASN* related OpenSSL methods under the hood. But there are other stuff besides ASN, like BIO - I'm trying to make a list.

**Comment by bartonjs** [comment]
> I copy-pasted some C# example for RSA and it contained CryptoConfig.MapNameToOID("SHA1") which uses two ASN* related OpenSSL methods under the hood.

That'd be an old sample, since the modern API would use `HashAlgorithmName.SHA1` instead of CryptoConfig.  That said, OID dotted-decimal to/from friendly-name mapping is done by a centralized class that has well-known values hard-coded.  Basically, Android would want something like the OSX version of the resolver, which is "the system doesn't have a function I can call for this, so only use the hard coded entries": https://github.com/dotnet/runtime/blob/master/src/libraries/System.Security.Cryptography.Encoding/src/Internal/Cryptography/OidLookup.OSX.cs

**Comment by grendello** [comment]
> My concern would be the potential of playing whack-a-mole with every release. Or if they somehow make system OpenSSL completely inaccessible.

Also, `dlopen(NULL)` segfaults in some versions of bionic.

**Comment by steveisok** [comment]
> > My concern would be the potential of playing whack-a-mole with every release. Or if they somehow make system OpenSSL completely inaccessible.
> 
> Also, `dlopen(NULL)` segfaults in some versions of bionic.

@grendello Do you know what version(s)?

**Comment by grendello** [comment]
> > > My concern would be the potential of playing whack-a-mole with every release. Or if they somehow make system OpenSSL completely inaccessible.
> > 
> > 
> > Also, `dlopen(NULL)` segfaults in some versions of bionic.
> 
> @grendello Do you know what version(s)?

It was discovered [in 2012](https://github.com/mono/mono/commit/9b73cd9c33507ba27829728ea82d10d98335e1d6) and bionic appears to have gotten dynamic loader fixes (e.g. [DT_SONAME](https://github.com/xamarin/xamarin-android/commit/6e6d3203d841a7b896ec37438e395f7f627d3cae) support) only relatively recently so I don't really know the versions.  There might be millions of Android devices with this problem, we simply don't know.

**Comment by EgorBo** [comment]
Closing this one since `dlopen(NULL)` is not relieable.

**Pull request closed without merging**